src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2009-2016 The Broad Institute
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package htsjdk.samtools.reference;
+
+import htsjdk.samtools.util.IterableOnceIterator;
+import htsjdk.samtools.util.SamLocusIterator;
+
+import java.util.List;
+
+import static htsjdk.samtools.util.SamLocusIterator.*;
+
+/**
+ * Iterator that traverses a SAM File and a ReferenceFile, accumulating information on a per-locus basis.
+ * Only, whatever loci are covered by the input reads are returned.
+ * Duplicate reads and non-primary alignments are filtered out.
+ * Iterator element holds both pileup (in the form of a LocusInfo object) and the reference base
+ *
+ * @author Yossi Farjoun
+ */
+public class SamLocusAndReferenceIterator extends IterableOnceIterator<SamLocusAndReferenceIterator.SAMLocusAndReference> {
+
+    private final ReferenceSequenceFileWalker referenceSequenceFileWalker;
+    private final SamLocusIterator locusIterator;
+
+    public SamLocusAndReferenceIterator(final ReferenceSequenceFileWalker referenceFile, final SamLocusIterator locusIterator) {
```

Comment:
Should there be any sort of check that the reference is consistent with the sam file?